### DISCUSSION TITLE
RCA Cambio W101 v2 firmware.fw Install

### DISCUSSION THREAD

**Discussion by cjohnson80** [onitake/gsl-firmware#193] [Q&A]
I'm struggling to get this driver working on this tablet. I feel like I'm missing something here.
I can modify the touchscreen_dmi.c but struggle with recompiling the kernel. Is there an up to date walkthrough or guide for this?

**Comment by onitake** [comment]
Which distribution are you using?
Are you getting compilation errors?

**Reply by cjohnson80** [reply]
I started with Fedora (which is my preferred distro) but then switched to ubuntu to try that. I currently have touch working but it seems as if the touch layer is rotated 90 degrees and there I'm trying to work through the problem.

**Reply by jwrdegoede** [reply]
> I started with Fedora (which is my preferred distro) but then switched to ubuntu to try that. I currently have touch working but it seems as if the touch layer is rotated 90 degrees and there I'm trying to work through the problem.

So you have been able to build your own kernel with a modified touchscreen-dmi.c ?

If yes, then you will want to add:

```
PROPERTY_ENTRY_BOOL("touchscreen-swapped-x-y"),
```

to the struct property_entry foo[] array you have made your DMI table entry point to, that will swap the axis, then you may still need to invert the coordinates on some of the axis and also get the min and size values correct. You can measure the min/max(aka size) values using e.g. evemu-record or evemu-test.

(if you still have not succeeded in building your own kernel please also let me know I have been working on a solution for that and might be able to help)

**Reply by cjohnson80** [reply]
I'm just getting setup to retry Fedora from scratch. I think I might know where the issue was but your other solution sounds interesting too. 
I'll setup to compile the kernel tonight and as it takes several hours and report back

**Reply by jwrdegoede** [reply]
> I'm just getting setup to retry Fedora from scratch.

Good, part of my other way for figuring out the right settings is switching back to Fedora.

I have been working on a set of kernel-patches which allows setting the same settings as which touchscreen-dmi does as a module option:

https://github.com/jwrdegoede/linux-sunxi/commit/d7b777f1d1dbbe76f61ce0d16f341fa5ecf26e8a
https://github.com/jwrdegoede/linux-sunxi/commit/c822b51039260030c24baa78dec8f225785a8a13

I have prepared a src.rpm of a modfied kernel and the plan was to get koji (the Fedora build system) to build this for you so you can just download the rpms, but atm there are some connection issues with koji which are blocking me from uploading the srpm. I will add another comment here when I have succeeded in starting a build.

What this kernel (once build and installed) will allow you to do is the following:

Step 1:

```
sudo rmmod silead
sudo modprobe silead
dmesg | tail
```

This will show the default firmware filename the silead driver expects, save your firmware under /lib/firmware/silead/... using that name. Then repeat step 1. now the error about the driver not being able to load the firmware should be gone:

Step 2:

```
sudo dnf install evtest
sudo evtest
```

And then select the touchscreen and run your fingers over the edges writing down the min/max x and y values.
These give you the touchscreen-min-x, touchscreen-min-y, touchscreen-size-x and touchscreen-size-y values
for your touchscreen. Note getting the min/max values is easier if you do it per axis like this:

```
sudo evtest /dev/input/event# | grep ABS_X
sudo evtest /dev/input/event# | grep ABS_Y
```

Replacing the `#` with the event device number for your touchscreen.

Step 3:

```
sudo rmmod silead
sudo modprobe silead settings=touchscreen-min-x=$(minx);touchscreen-min-y=$(miny);touchscreen-size-x=$(max);touchscreen-size-y=$(maxy);touchscreen-swapped-x-y
```

Replacing all the `$(var)` bits with min/max x/y values from Step 2. Note the driver does the axis swapping as the very last processing step, so the min/max x/y values should be those measure when not doing the axis swapping.

After this you may also need to invert/mirror one or both of the axis for everything to work. To do this simple repeat Step 3 appending `;touchscreen-inverted-y` and/or `;touchscreen-inverted-x` to the settings.

Step 4:

Your touchscreen may also have a capacitive home button (windows logo) outside the touch area. To test his repeat the rmmod + modprobe with `;silead,home-button` appended to the settings and then check in evtest if pressing the windows logo generates events. If the windows logo does not generate events then remove the  `;silead,home-button` from settings again.

Assuming you get everything to work, please let me know the final settings string you need; and attach the output of the dmidecode command here and then I can turn this into a touchscreen_dmi.c patch for you. Or you can write such a patch yourself once you have figured out the settings without needing to build a new kernel + reboot all the time to test.

Note you can already do Step 1 and Step 2 with the default Fedora kernel only step 3 requires the new kernel which adds the settings option to the silead driver.

**Reply by jwrdegoede** [reply]
A kernel with the patches is now building here:
https://koji.fedoraproject.org/koji/taskinfo?taskID=92396301

This should be finished in about an hour or so. See here for some generic instructions for installing a kernel from a koji build: https://fedorapeople.org/~jwrdegoede/kernel-test-instructions.txt

BTW it would be great if you can write down notes / improvements for my instructions from above to help improve them. I hope to get the patches adding the setting= module parameter upstream soon and then it would be good if we can get some well tested instructions up in the README here to help people get the touchscreen working on devices which are not supported yet.

**Reply by jwrdegoede** [reply]
Note the build has just finished, so go and grab your rpms while they are still fresh: https://koji.fedoraproject.org/koji/taskinfo?taskID=92396301

**Reply by cjohnson80** [reply]
I finished compiling but something still isn't right. Haven't had much time to troubleshoot yet but hopefully will get it working. 
I am, however, intrigued by your kernel. I'll give that a go and see what results I can get out of it too

**Reply by cjohnson80** [reply]
I started from scratch and tried out your kernel by installing as per your generic instructions, but I had no success.
The output from dmesg is this
[   79.394131] silead_ts i2c-MSSL1680:00: supply vddio not found, using dummy regulator
[   79.394940] silead_ts i2c-MSSL1680:00: supply avdd not found, using dummy regulator
[   79.395264] i2c_designware 80860F41:03: i2c_dw_handle_tx_abort: lost arbitration
[   80.457836] i2c_designware 80860F41:03: controller timed out
[   80.457857] silead_ts i2c-MSSL1680:00: Chip ID read error -110
[   80.458683] silead_ts: probe of i2c-MSSL1680:00 failed with error -110

**Reply by cjohnson80** [reply]
I thihnk I copied the wrong firmware into the /lib/firmware/silead folder. After making that change I have some touch happening.
I continued along the list with running evtest
Found the variables for min/max x & y
sudo rmmod silead
sudo modprobe silead settings=touchscreen-min-x=4;touchscreen-min-y=20;touchscreen-size-x=1644;touchscreen-size-y=874;touchscreen-swapped-x-y

This is the output as follows:

bash: touchscreen-min-y=20: command not found...
bash: touchscreen-size-x=1644: command not found...
bash: touchscreen-size-y=874: command not found...
bash: touchscreen-swapped-x-y: command not found...

**Reply by jwrdegoede** [reply]
Ah, my bad. (ba)sh thinks the commadline ends at the first `;` so you need to put single or double quotes around the entire settings= argument. Please give that a try.

**Reply by cjohnson80** [reply]
Ok cool. That did the trick! (I gave it a quick go before heading out) Everything seems to be working as expected touch wise but I’ll have to give it some further testing when I get back home. 
Then I’ll see if I can duplicate it again.

**Comment by cjohnson80** [comment]
Here are the settings I used for my device: modprobe silead settings="touchscreen-min-x=4;touchscreen-min-y=20;touchscreen-size-x=1644;touchscreen-size-y=874;touchscreen-swapped-x-y"
What information do you need from dmidecode? or just everything?

**Reply by jwrdegoede** [reply]
Just everything is easiest. Often some of the DMI strings are quite generic because the vendor just left the template intact, so we get DMI strings like "Default String". So I always look through a bunch of options for a unique combination to create the quirk.

**Reply by cjohnson80** [reply]
Here is the full output
[dmidecode RCA Cambio.txt](https://github.com/onitake/gsl-firmware/files/9694783/dmidecode.RCA.Cambio.txt)

**Reply by jwrdegoede** [reply]
Hmm, that DMI info is surprisingly clean/useful (good). Attached here is the patch which I have cooked up. Note when running a kernel with this patch, the kernel will expect the firmware file to be named `/lib/firmware/silead/gsl1680-rca-cambio-w101-v2.fw` so please make a copy of your firmware there.

[0001-platform-x86-touchscreen_dmi-Add-info-for-the-RCA-Ca.patch.txt](https://github.com/onitake/gsl-firmware/files/9697085/0001-platform-x86-touchscreen_dmi-Add-info-for-the-RCA-Ca.patch.txt)

A Fedora kernel with this patch added is now building here:

https://koji.fedoraproject.org/koji/taskinfo?taskID=92563362

(finished in about 2 hours from now I guess)

Please give this a try. If this work I will push the patch to the official mainline Linux kernels. If this works can you please submit a pull-request here, adding your firmware as `firmware/linux/silead/gsl1680-rca-cambio-w101-v2.fw` ?

**Reply by jwrdegoede** [reply]
p.s. Note the new kernel no longer has the settings= patch ..., so if it does not work, don't expect the settings= thing to work with this new kernel build.

**Reply by cjohnson80** [reply]
Not sure if I'm doing something wrong but this doesn't seem to work.

**Reply by jwrdegoede** [reply]
What do `uname -a` and `dmesg | grep silead` give as output (when run from a terminal)  ?

**Comment by cjohnson80** [comment]
uname -a && dmesg | grep silead
Linux fedora 5.19.12-200.gsl_fw_dmi.fc36.x86_64 #1 SMP PREEMPT_DYNAMIC Mon Oct 3 12:24:20 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
[   15.236761] silead_ts i2c-MSSL1680:00: supply vddio not found, using dummy regulator
[   15.237102] silead_ts i2c-MSSL1680:00: supply avdd not found, using dummy regulator
[   15.239780] silead_ts i2c-MSSL1680:00: Silead chip ID: 0x50910000
[   16.831876] input: silead_ts as /devices/platform/80860F41:03/i2c-3/i2c-MSSL1680:00/input/input10

**Reply by jwrdegoede** [reply]
Hmm, what does evtest show as output now ? What are the value of the ranges of the axis reported in evtest  ?

**Reply by cjohnson80** [reply]
$ sudo evtest /dev/input/event10
Input driver version is 1.0.1
Input device ID: bus 0x19 vendor 0x1 product 0x1 version 0x100
Input device name: "gpio-keys"
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 114 (KEY_VOLUMEDOWN)
    Event code 115 (KEY_VOLUMEUP)
Key repeat handling:
  Repeat type 20 (EV_REP)
    Repeat code 0 (REP_DELAY)
      Value    250
    Repeat code 1 (REP_PERIOD)
      Value     33
Properties:
Testing ... (interrupt to exit)

**Reply by jwrdegoede** [reply]
In case you have not noticed yourself, it looks like /dev/input/event10 is no longer the touchscreen. The event# numbers are not stable. Probably easiest to run `sudo evtest` without an extra argument and then pick the right number from the list.

**Reply by cjohnson80** [reply]
I did notice that but was just testing you ;) 
dmesg has it as event10 but using evtest it is found as event9 and here is the output:
Input driver version is 1.0.1
Input device ID: bus 0x18 vendor 0x0 product 0x0 version 0x0
Input device name: "silead_ts"
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 330 (BTN_TOUCH)
  Event type 3 (EV_ABS)
    Event code 0 (ABS_X)
      Value      0
      Min       20
      Max      873
    Event code 1 (ABS_Y)
      Value      0
      Min        4
      Max     1643
    Event code 47 (ABS_MT_SLOT)
      Value      0
      Min        0
      Max        9
    Event code 53 (ABS_MT_POSITION_X)
      Value      0
      Min       20
      Max      873
    Event code 54 (ABS_MT_POSITION_Y)
      Value      0
      Min        4
      Max     1643
    Event code 57 (ABS_MT_TRACKING_ID)
      Value      0
      Min        0
      Max    65535
Properties:
  Property type 1 (INPUT_PROP_DIRECT)
Testing ... (interrupt to exit)

**Reply by jwrdegoede** [reply]
So the min-max values there do show that the DMI quirk which I added is being applied properly...

Do you get events in evtest when trying to use the touchscreen ?

I wonder if you maybe copied the wrong fw file to the fw filename used by the kernel with the DMI quirk ?

**Reply by cjohnson80** [reply]
There are no touch screen events. Let me see if the firmware is the issue

**Reply by cjohnson80** [reply]
The firmware seems to be the correct one. It's the only one that loads

**Reply by jwrdegoede** [reply]
Weird.

Some things to try:

1. rmmod + modprobe silead, like you did when passing the settings
2. powercycle the device
3. try the previous kernel where things did work with the settings=..  modprobe argument

The DMI patch I added merely sets the parameters you were passing manually, so there is something else going on here (unrelated to the patches) which is causing the touchscreen to not work.

**Reply by cjohnson80** [reply]
I did a reinstall to try from a vanilla state
here is the output from dmesg and evtest
[chris@fedora ~]$ dmesg | grep silead
[   14.747199] silead_ts i2c-MSSL1680:00: supply vddio not found, using dummy regulator
[   14.750344] silead_ts i2c-MSSL1680:00: supply avdd not found, using dummy regulator
[   14.750741] silead_ts i2c-MSSL1680:00: Silead chip ID: 0x50910000
[   16.211011] input: silead_ts as /devices/platform/80860F41:03/i2c-3/i2c-MSSL1680:00/input/input9
[chris@fedora ~]$ sudo evtest
[sudo] password for chris: 
No device specified, trying to scan all of /dev/input/event*
Available devices:
/dev/input/event0:	Windows USB Keyboard
/dev/input/event1:	Windows USB Keyboard System Control
/dev/input/event2:	Windows USB Keyboard Consumer Control
/dev/input/event3:	Windows USB Keyboard Mouse
/dev/input/event4:	Video Bus
/dev/input/event5:	Intel HDMI/DP LPE Audio HDMI/DP,pcm=0
/dev/input/event6:	Intel HDMI/DP LPE Audio HDMI/DP,pcm=1
/dev/input/event7:	axp20x-pek
/dev/input/event8:	silead_ts
/dev/input/event9:	gpio-keys
/dev/input/event10:	gpio-keys
/dev/input/event11:	Bluetooth Mouse
/dev/input/event12:	Bluetooth Mouse
Select the device event number [0-12]:

**Reply by jwrdegoede** [reply]
That evtest output looks incomplete ?  Also did the re-install help ?

**Reply by cjohnson80** [reply]
It didn't help. still the same results. running evtest and selecting the event* for silead_ts produces no touch feedback either.

**Reply by jwrdegoede** [reply]
Did you try the rmmod + modprobe thing?  That will execute the ACPI _PS3 (off) + _PS0 (on) methods, which may be what did the trick last time ...

**Reply by cjohnson80** [reply]
Yeah here are the results:
[chris@fedora ~]$ sudo rmmod silead
[sudo] password for chris: 
[chris@fedora ~]$ sudo modprobe silead
[chris@fedora ~]$ dmesg | tail
[11105.928212] perf: interrupt took too long (6422 > 6405), lowering kernel.perf_event_max_sample_rate to 31000
[16862.062175] perf: interrupt took too long (8037 > 8027), lowering kernel.perf_event_max_sample_rate to 24000
[24319.161535] hid-generic 0005:3938:1001.0006: unknown main item tag 0x0
[24319.162042] input: Bluetooth Mouse as /devices/platform/80860F0A:00/serial0/serial0-0/bluetooth/hci0/hci0:2/0005:3938:1001.0006/input/input20
[24319.163261] input: Bluetooth Mouse as /devices/platform/80860F0A:00/serial0/serial0-0/bluetooth/hci0/hci0:2/0005:3938:1001.0006/input/input22
[24319.163576] hid-generic 0005:3938:1001.0006: input,hidraw2: BLUETOOTH HID v0.1e Mouse [Bluetooth Mouse] on 38:a2:8c:65:a9:63
[26690.811502] silead_ts i2c-MSSL1680:00: supply vddio not found, using dummy regulator
[26690.812428] silead_ts i2c-MSSL1680:00: supply avdd not found, using dummy regulator
[26690.813234] silead_ts i2c-MSSL1680:00: Silead chip ID: 0x50910000
[26693.963134] input: silead_ts as /devices/platform/80860F41:03/i2c-3/i2c-MSSL1680:00/input/input23
[chris@fedora ~]$ sudo evtest
No device specified, trying to scan all of /dev/input/event*
Available devices:
/dev/input/event0:	Windows USB Keyboard
/dev/input/event1:	Windows USB Keyboard System Control
/dev/input/event2:	Windows USB Keyboard Consumer Control
/dev/input/event3:	Windows USB Keyboard Mouse
/dev/input/event4:	Video Bus
/dev/input/event5:	Intel HDMI/DP LPE Audio HDMI/DP,pcm=0
/dev/input/event6:	Intel HDMI/DP LPE Audio HDMI/DP,pcm=1
/dev/input/event7:	axp20x-pek
/dev/input/event8:	silead_ts
/dev/input/event9:	gpio-keys
/dev/input/event10:	gpio-keys
/dev/input/event11:	Bluetooth Mouse
/dev/input/event12:	Bluetooth Mouse
Select the device event number [0-12]: 8
Input driver version is 1.0.1
Input device ID: bus 0x18 vendor 0x0 product 0x0 version 0x0
Input device name: "silead_ts"
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 330 (BTN_TOUCH)
  Event type 3 (EV_ABS)
    Event code 0 (ABS_X)
      Value      0
      Min       20
      Max      873
    Event code 1 (ABS_Y)
      Value      0
      Min        4
      Max     1643
    Event code 47 (ABS_MT_SLOT)
      Value      0
      Min        0
      Max        9
    Event code 53 (ABS_MT_POSITION_X)
      Value      0
      Min       20
      Max      873
    Event code 54 (ABS_MT_POSITION_Y)
      Value      0
      Min        4
      Max     1643
    Event code 57 (ABS_MT_TRACKING_ID)
      Value      0
      Min        0
      Max    65535
Properties:
  Property type 1 (INPUT_PROP_DIRECT)
Testing ... (interrupt to exit)

**Reply by jwrdegoede** [reply]
Did you try to see if touches now get registered or not ?

If touches still do not get registered, what happens if you install my previous kernel from: https://koji.fedoraproject.org/koji/taskinfo?taskID=92396301 and then do rmmod + "modprobe silead settings=..." again. IOW go back to what was a known working combination ?

**Comment by cjohnson80** [comment]
There are no touches after that. I'll try going back to that kernel

**Reply by cjohnson80** [reply]
Everything works as expected with the first kernel using the settings option

**Reply by cjohnson80** [reply]
So I recompiled a 6.0.0 kernel with your patch info and it seems to work.  A couple of times it didn't work on reboot but rmmod and modprobe silead fixed that. I'll test it further

**Reply by jwrdegoede** [reply]
That is good to hear, thank you for testing kernel 6.0.0 with the patch.

Can you try if adding this change to touchscreen_dmi.c helps with things sometimes not working on reboot ? :

```
diff --git a/drivers/platform/x86/touchscreen_dmi.c b/drivers/platform/x86/touchscreen_dmi.c
index baae3120efd0..2fede89e6390 100644
--- a/drivers/platform/x86/touchscreen_dmi.c
+++ b/drivers/platform/x86/touchscreen_dmi.c
@@ -776,6 +776,7 @@ static const struct property_entry rca_cambio_w101_v2_props[] = {
 	PROPERTY_ENTRY_U32("touchscreen-size-x", 1644),
 	PROPERTY_ENTRY_U32("touchscreen-size-y", 874),
 	PROPERTY_ENTRY_BOOL("touchscreen-swapped-x-y"),
+	PROPERTY_ENTRY_BOOL("silead,stuck-controller-bug"),
 	PROPERTY_ENTRY_STRING("firmware-name", "gsl1680-rca-cambio-w101-v2.fw"),
 	PROPERTY_ENTRY_U32("silead,max-fingers", 10),
 	{ }
```

This will make the driver (try to) power-cycle the controller at probe time, effectively emulating the rmmod + modprobe results.

Also can you run: `sudo acpidump -o acpidump.hex` and then attach the generated acpidump.hex file here ?  Then I can check if the touchscreen part of your ACPI tables is doing anything weird.

**Reply by cjohnson80** [reply]
Here is a copy of the acpidump
[acpidump.hex.txt](https://github.com/onitake/gsl-firmware/files/9766728/acpidump.hex.txt)

**Reply by jwrdegoede** [reply]
Thank you for the acpidump. The ACPI tables do implement a _PS0 method resetting the controller. Can you try adding `PROPERTY_ENTRY_BOOL("silead,stuck-controller-bug"),` to `rca_cambio_w101_v2_props[]` in drivers/platform/x86/touchscreen_dmi.c and see if that helps with the touchscreen sometimes not working at boot ?

Note you only need to run `make bzImage` and then copy `arch/x86/boot/bzImage /boot/vmlinuz-...` no need to entirely rebuild the kernel after this small change.

**Reply by cjohnson80** [reply]
I tried to add the change to the dmi.c but I still don't get any results. i still seem to have to rmmod and modprobe to get touch to work

**Reply by jwrdegoede** [reply]
Did you rebuild the vmlinuz `make bzImage` and then copy `arch/x86/boot/vmlinuz` to `/boot/vmlinuz-<version>` ? The touchscreen_dmi.c file gets build into the main kernel image, it is not a module.

**Reply by cjohnson80** [reply]
Yes I did that but still don’t have results. I did a few fresh kernel rebuilds too just to be sure

**Reply by jwrdegoede** [reply]
Weird, I'm afraid I'm all out of ideas then.

**Reply by jwrdegoede** [reply]
I'll go and submit the touchscreen_dmi.c patch upstream as is then, since the addition of the `"silead,stuck-controller-bug"` property does not help.

**Reply by cjohnson80** [reply]
Yeah I don’t think it would be a bad idea. I’m still playing around with different variations and recompiling kernels. I’ll post here if I happen to figure something out.

**Comment by chrisma82** [comment]
Unrelated sorry but can anyone provide some driver files from this device? Rca support is not helpful. Thanks!
